### PR TITLE
Uniswap V3 Backtesting 

### PR DESCRIPTION
**Pull request by AlexTheLion123**
Adds support for uniswap V3 backtesting + a jupyter notebook that does a backtesting on Uni V3. 

This is done by:

- Adding Uniswap V3 compatibility to `routing_data.py`
- Adding a new file `uniswap_v3_routing.py`
- Adding Uniswap V3 options to `default_routing_options.py`
- Adding `notebooks/ema_crossover_uniswap_v3.ipynb`

`routing_data.py` changes:

- Created `validate_reserve_currency` method
- Created `get_uniswap_v2_compatible_routing_types` and `get_uniswap_v2_compatible_routing_types` methods to easily check which type of routing type is required
- Black formatting was run on `routing_data.py`, so there are fairly extensive formatting changes as well.

### CODE REVIEW

**Comment by AlexTheLion123** [comment]
fixes #140, fixes #188

File: tradeexecutor/ethereum/uniswap_v3_routing.py [hunk 1]
```diff
@@ -0,0 +1,691 @@
...
+    translate_token
+from tradingstrategy.pair import PandasPairUniverse
+
+from tradeexecutor.strategy.universe_model import StrategyExecutionUniverse
+
+
+logger = logging.getLogger(__name__)
+
+
+class OutOfBalance(Exception):
```
Comment (miohtama) [inline tradeexecutor/ethereum/uniswap_v3_routing.py:39]: Use same exception for Uniswap v2 and v3

File: tradeexecutor/ethereum/uniswap_v3_routing.py [hunk 2]
```diff
@@ -0,0 +1,691 @@
...
+    """Did not have enough tokens"""
+
+
+class UniswapV3RoutingState(RoutingState):
```
Comment (miohtama) [inline tradeexecutor/ethereum/uniswap_v3_routing.py:43]: Use same base class for v2 and v3 routing, so that we do not copy paste any code

**Review by miohtama — CHANGES_REQUESTED** [review]

**Comment by AlexTheLion123** [comment]
@miohtama Have created `RoutingState` base class in `routing_state.py`

**Review by miohtama — CHANGES_REQUESTED** [review]

File: tradeexecutor/ethereum/uniswap_v2_routing.py [hunk 3]
```diff
@@ -1,193 +1,64 @@
...
-    - Swap tx creation
-
-    Manage the state of already given approvals here,
-    so that we do not do duplicates.
-
-    The approvals are not persistent in the executor state,
-    but are specific for each cycle.
-    """
-
+class UniswapV2RoutingState(RoutingStateBase):
```
[the remark this answers is not in the document]
Comment (AlexTheLion123) [reply tradeexecutor/ethereum/uniswap_v2_routing.py:34]: @miohtama It isn't the same, there is `UniswapV3RoutingState(RoutingStateBase)` as well. 
https://github.com/tradingstrategy-ai/trade-executor/blob/52e38e009110c53f184dabea115f7d146e1725c2/tradeexecutor/ethereum/uniswap_v3_routing.py#L34-L117

[the remark this answers is not in the document]
Comment (AlexTheLion123) [reply tradeexecutor/ethereum/uniswap_v2_routing.py:34]: @miohtama It isn't the same, there is `UniswapV3RoutingState(RoutingStateBase)` as well. 
https://github.com/tradingstrategy-ai/trade-executor/blob/52e38e009110c53f184dabea115f7d146e1725c2/tradeexecutor/ethereum/uniswap_v3_routing.py#L34-L117

File: tradeexecutor/ethereum/uniswap_v3_routing.py [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+            quote_token=quote_token,
+            pool_fees=pool_fees,
+            amount_in=reserve_amount,
+            max_slippage=max_slippage * 100,  # In BPS,
+            intermediate_token=intermediary_token,
+        )
+        
+        return self.get_signed_tx(bound_swap_func, self.swap_gas_limit)
+
+class UniswapV3SimpleRoutingModel(RoutingModel):
```
[the remark this answers is not in the document]
Comment (AlexTheLion123) [reply tradeexecutor/ethereum/uniswap_v3_routing.py:119]: Will add a base class for the routing model as well.

**Review by miohtama — CHANGES_REQUESTED** [review]
- Still need a test case that uses the new Uni v3 classes

**Pull request merged**